addon/components/base/bs-form.js

```diff
@@ -220,26 +246,40 @@ export default Component.extend({
    * @private
    */
   submit(e) {
+    this.set('isSubmitting', true);
+
     if (e) {
       e.preventDefault();
     }
     let model = this.get('model');
 
     this.get('onBefore')(model);
 
-    if (!this.get('hasValidator')) {
-      return this.get('onSubmit')(model);
-    } else {
-      let validationPromise = this.validate(this.get('model'));
-      if (validationPromise && validationPromise instanceof RSVP.Promise) {
-        validationPromise
-          .then((r) => this.get('onSubmit')(model, r))
-          .catch((err) => {
-            this.set('showAllValidations', true);
-            return this.get('onInvalid')(model, err);
-          });
-      }
+    let validationPromise = this.get('hasValidator') ? this.validate(this.get('model')) : RSVP.resolve();
+    if (!isPromise(validationPromise)) {
+      validationPromise = RSVP.resolve();
     }
+    validationPromise
+      .then((r) => {
+        let onSubmitPromise = this.get('onSubmit')(model, r);
+        if (!isPromise(onSubmitPromise)) {
+          onSubmitPromise = RSVP.resolve();
+        }
+        onSubmitPromise
+          .finally(() => {
```

Comment:
I think we must be careful with `finally()`. It is available when RSVP is used, and has been introduced to the ES-spec lately IIRC, but there still can be native Promise implementation around that don't support it. 

maybe we can solve this by restructuring the code above a bit, to always start with a `RSVP.resolve()` and call `.then(()  => this.get('onSubmit')(model, r))` (doesn't matter if it return a promise now or not). Now it *should* always be a RSVP promise which supports `finally()`, right?